### PR TITLE
Remove usage of default HTTP client and transport

### LINKED ISSUE #2533: Remove usage of default HTTP Client or Transport
#### Summary

We should stop using the default HTTP client, and instead use one we get from the `component.(*Component)` so that we can pass it TLS client config (such as trusted CAs).

#### What is already there? What do you see now?

Some places still use `http.DefaultClient`

#### What is missing? What do you want to see?

We should use an http Client that we can configure

#### How do you propose to implement this?

- Add a `func (c *Component) HTTPTransport() http.RoundTripper`
- Add a `func (c *Component) HTTPClient() *http.Client`
- Use those in places where we make HTTP calls
	- `pkg/applicationserver/io/packages/loradms/v1`
	- `pkg/applicationserver`
	- `pkg/fetch`
	- `pkg/identityserver/picture`
	- `pkg/interop`
	- `pkg/web/oauthclient`
	- maybe other places

### PR DESCRIPTION
**Pull request by neoaggelos**
#### Summary

Closes #2533

#### Changes

- Adds `component.HTTPClient(context.Context) (*http.Client, error)` and `component.HTTPTransport(context.Context) (http.RoundTripper, error)`
- Extend `fetcher.FromHTTP` to accept a custom transport.
- Add new configuration `Transport http.RoundTripper` for all fetchers.
- (as) Extend `io.Server` with `HTTPClient()` method and use it for webhooks and the LoRaCloud package.
- (oauthclient) Use the component HTTPClient.

#### Testing

Unit tests, test locally.

##### Regressions

Possible, since different parts of the codebase dealt with HTTP clients and transports in all sorts of ways.

#### Notes for Reviewers

Please make sure your code-owned files look OK.

@johanstokking `pkg/interop` client intentionally not changed, since it would require quite some refactoring which I would prefer to do in a separate PR.
@kschiffer specifically, what is a scenario to verify that `oauthclient` works as expected?

#### Checklist

- [X] Scope: The referenced issue is addressed, there are no unrelated changes.
- [X] Compatibility: The changes are backwards compatible with existing API, storage, configuration and CLI, according to the compatibility commitments in `README.md` for the chosen target branch.
- [ ] Documentation: Relevant documentation is added or updated.
- [X] Changelog: Significant features, behavior changes, deprecations and fixes are added to `CHANGELOG.md`.
- [X] Commits: Commit messages follow guidelines in `CONTRIBUTING.md`, there are no fixup commits left.

### CODE REVIEW

File: cmd/ttn-lw-stack/commands/start.go
Comment (johanstokking) [inline cmd/ttn-lw-stack/commands/start.go]: Should we continue execution?

File: cmd/ttn-lw-stack/commands/start.go
Comment (johanstokking) [inline cmd/ttn-lw-stack/commands/start.go]: I think it's quite error prone to set transport settings here.

Can we were this in the component (context)?

File: pkg/applicationserver/config.go
Comment (johanstokking) [inline pkg/applicationserver/config.go]: How about returning the error?

File: pkg/applicationserver/io/packages/loradms/v1/package.go
Comment (johanstokking) [inline pkg/applicationserver/io/packages/loradms/v1/package.go]: Idem

**Review by johanstokking — DISMISSED** [review]
Stopped review halfway. Please rework and mark as ready when done

File: pkg/applicationserver/io/packages/loradms/v1/package.go
Comment (johanstokking) [reply pkg/applicationserver/io/packages/loradms/v1/package.go]: What do you mean with "tested locally" if you didn't see this panicking?

File: pkg/component/http.go
Comment (adriansmares) [inline pkg/component/http.go]: Not sure if I like this.

First of all, this becomes `component.WithTimeout`, which conflicts with actual component options. 
Second, I think it's a bit of an anti-pattern to override this everywhere - We added `HTTPClient` in order to have uniform HTTP client settings everywhere. I'd prefer that this option is used only in tests at most.

File: pkg/component/http.go
Comment (adriansmares) [inline pkg/component/http.go]: Error handling for `err` here is missing.

File: pkg/component/http.go
Comment (adriansmares) [inline pkg/component/http.go]: Return the error, don't hide it.

**Review by adriansmares — CHANGES_REQUESTED** [review]

**Review by kschiffer — COMMENTED** [review]
> @kschiffer specifically, what is a scenario to verify that oauthclient works as expected?

Testing whether client authorization e.g. of the Console still works. I tested this both with and without TLS and it worked. Codeowned files also LGTM.

File: pkg/component/http.go
Comment (htdvisser) [reply pkg/component/http.go]: We should probably also check how this interacts with the context deadline that we use in individual requests.

File: pkg/component/http.go
Comment (htdvisser) [reply pkg/component/http.go]: Technically not missing, because it is returned in the end.

But I would indeed prever doing error handling here, and returning `client, nil` in the end of the func.

File: pkg/component/http.go
Comment (htdvisser) [reply pkg/component/http.go]: Technically not hidden, because it is returned in the end.

But I would indeed prever doing error handling here, and returning `transport, nil` in the end of the func.

File: pkg/fetch/http.go
Comment (htdvisser) [inline pkg/fetch/http.go]: Shouldn't this be above the `if cache {`?

Comment (neoaggelos) [reply pkg/fetch/http.go]: It shouldn't make a difference, since `httpcache.Transport` will use `http.DefaultTransport` if transport is `nil`. I can see how that may be confusing though, so I moved it before the `if cache {` block.

File: cmd/ttn-lw-stack/commands/start.go
Comment (adriansmares) [reply cmd/ttn-lw-stack/commands/start.go]: This should indeed halt the execution, please `return err`.

File: pkg/applicationserver/io/packages/loradms/v1/package.go
Comment (adriansmares) [inline pkg/applicationserver/io/packages/loradms/v1/package.go]: Remove the diff (the `()`).

File: pkg/component/http.go
Comment (adriansmares) [inline pkg/component/http.go]: Nit: You may remove the parenthesis.

**Review by adriansmares — APPROVED** [review]
I like this a lot more now.

File: cmd/ttn-lw-stack/commands/start.go
Comment (rvolosatovs) [inline cmd/ttn-lw-stack/commands/start.go]:
```suggestion
		if conf := config.ServiceBase.DeviceRepository; conf.Transport == nil {
```
https://github.com/TheThingsNetwork/lorawan-stack/blob/v3.10/DEVELOPMENT.md#single-word-entities

File: pkg/component/http.go
Comment (rvolosatovs) [inline pkg/component/http.go]: nit;
```suggestion
	return &http.Client{
		Timeout:   defaultHTTPClientTimeout,
		Transport: transport,
	}, nil
```

File: pkg/component/http.go
Comment (rvolosatovs) [inline pkg/component/http.go]: nit;
```suggestion
	tr, err := c.HTTPTransport(ctx)
```

File: pkg/component/http.go
Comment (rvolosatovs) [inline pkg/component/http.go]: nit; this is not used before L46 and it does not fail - do this on L45

File: pkg/networkserver/networkserver.go
Comment (rvolosatovs) [inline pkg/networkserver/networkserver.go]:
```suggestion
	var interopCl InteropClient
	if !conf.Interop.IsZero() {
		interopConf := conf.Interop
		interopConf.GetFallbackTLSConfig = func(ctx context.Context) (*tls.Config, error) {
			return c.GetTLSClientConfig(ctx)
		}
		interopConf.BlobConfig = c.GetBaseConfig(ctx).Blob
		if interopConf.Transport == nil {
			tr, err := c.HTTPTransport(ctx)
			if err != nil {
				return nil, err
			}
			interopConf.Transport = tr
		}
```

**Review by rvolosatovs — CHANGES_REQUESTED** [review]

**Review by rvolosatovs — APPROVED** [review]

**Comment by neoaggelos** [comment]
This needs codeowner approval to merge.

File: pkg/applicationserver/io/packages/loradms/v1/package.go
Comment (KrishnaIyer) [inline pkg/applicationserver/io/packages/loradms/v1/package.go]: This isn't related right?

**Review by KrishnaIyer — COMMENTED** [review]
Code owned files LGTM.

To what extent is this tested?

> Regressions
Possible, since different parts of the codebase dealt with HTTP clients and transports in all sorts of ways.

That doesn't seem very convincing to me.

**Comment by neoaggelos** [comment]
> > Regressions
> > Possible, since different parts of the codebase dealt with HTTP clients and transports in all sorts of ways.
> 
> That doesn't seem very convincing to me.

Indeed, most of this is not covered by our unit tests (but should be covered by the end to end tests). I'd rather this is merged early for 3.12 instead, to avoid disruptions with 3.11

**Review by adriansmares — APPROVED** [review]

**Review by KrishnaIyer — APPROVED** [review]
Code owned files LGTM

**Review by htdvisser — APPROVED** [review]

**Pull request merged**
